### PR TITLE
Add default worker thread pool

### PR DESCRIPTION
**Pull request by pierricgimmig**
Add ThreadPool::GetDefaultThreadPool() which returns a pointer to a
global instance of a fixed-sized thread pool. This is done in order to
minimize the boilerplate code of having to pass in a ThreadPool object
to any object wanting to parallelize tasks. It is similar in spirit to std::async
being available from anywhere, and to Qt's QThreadPool::globalInstance(),
and to our internal executor's default thread pool.

A call to "ThreadPool::InitializeDefaultThreadPool()" will create the default 
thread pool. We can also override the default thread pool by calling 
"ThreadPool::SetDefaultThreadPool(...)". Note that if neither of these
functions are called, the default thread pool will be created on the first 
call to "ThreadPool::GetDefaultThreadPool()".

Fix the "ThreadPool" destruction sequence that required a manual call to 
"ShutdownAndWait()", which is now called from the destructor.

Having access to this global thread pool simplifies the use of the "TaskGroup"
object, which now uses the default thread pool if no executor is specified.

b/225063353

Tests: FunctionsDataView filtering, which uses a TaskGroup works as
expected.

### CODE REVIEW

**Review by florian-kuebler — APPROVED** [review]
So to me that looks good. Thanks. However, I would like you to wait for another review.

**Comment by beckerhe** [comment]
I'm very much in favor of adding the ShutdownAndWait logic to the destructor. But I don't think having a global thread pool is a good idea at all. This is exactly the kind of coupling we are trying to fight. It will make unit tests harder because tasks might implicilty leak from one test into the next. Also lifetime is not properly preserved anymore. I believe the mentioned pointers to ThreadPool or Executor are a good thing. I would prefer to have them on each constructor that needs an executor.

I would love to hear what others think.

**Comment by florian-kuebler** [comment]
> I'm very much in favor of adding the ShutdownAndWait logic to the destructor. But I don't think having a global thread pool is a good idea at all. This is exactly the kind of coupling we are trying to fight. It will make unit tests harder because tasks might implicilty leak from one test into the next. Also lifetime is not properly preserved anymore. I believe the mentioned pointers to ThreadPool or Executor are a good thing. I would prefer to have them on each constructor that needs an executor.
> 
> I would love to hear what others think.

Good point. When reviewing this, I was already thinking of the tests, but obviously did not realized the issue that you mentioned. I think I am with you here.

**Comment by pierricgimmig** [comment]
> It will make unit tests harder because tasks might implicilty leak from one test into the next.

Which test in particular are you worried about? The existing ThreadPool tests will work as before. The TaskGroup tests will also work. If anything, isn't this making sure the tests are more robust by not having implicit assumptions about the state of the ThreadPool?

This change is about reducing friction. It's giving an option to use a default thread pool but is not forcing anyone to use it, we can always inject dependencies where it makes sense. I feel that the current state discourages parallelization in the code base because the boilerplate is cumbersome. Imagine if std::async or parallel-for loops required passing in a dependency... Also, with the current state, we're promoting the creation of ThreadPool objects, those are not cheap. Furthermore, for jobs, since we are constrained by the number of cores, it seems more efficient to have a single well defined fixed-sized ThreadPool.

Do you see any other way to simplify parallel execution? I really feel that passing in executors in every constructor is overkill. I would prefer not using std parallel algorithms as I think we have more control using our own executors, but at least they would be available from anywhere without having to touch multiple classes in order to pass in a dependency through a hierarchy to have access to a simple parallel-for loop.

Btw, Qt also seems to agree that a global thread pool is useful. QThreadPool::globalInstance() is the default thread pool for the QtConcurrent algorithms: https://doc.qt.io/qt-5/qthreadpool.html, https://doc.qt.io/qt-5/qtconcurrent.html.

**Comment by beckerhe** [comment]
> > It will make unit tests harder because tasks might implicilty leak from one test into the next.
> 
> Which test in particular are you worried about? The existing ThreadPool tests will work as before. The TaskGroup tests will also work. If anything, isn't this making sure the tests are more robust by not having implicit assumptions about the state of the ThreadPool?
> 
> This change is about reducing friction. It's giving an option to use a default thread pool but is not forcing anyone to use it, we can always inject dependencies where it makes sense. I feel that the current state discourages parallelization in the code base because the boilerplate is cumbersome. Imagine if std::async or parallel-for loops required passing in a dependency... Also, with the current state, we're promoting the creation of ThreadPool objects, those are not cheap. Furthermore, for jobs, since we are constrained by the number of cores, it seems more efficient to have a single well defined fixed-sized ThreadPool.
> 
> Do you see any other way to simplify parallel execution? I really feel that passing in executors in every constructor is overkill. I would prefer not using std parallel algorithms as I think we have more control using our own executors, but at least they would be available from anywhere without having to touch multiple classes in order to pass in a dependency through a hierarchy to have access to a simple parallel-for loop.
> 
> Btw, Qt also seems to agree that a global thread pool is useful. QThreadPool::globalInstance() is the default thread pool for the QtConcurrent algorithms: https://doc.qt.io/qt-5/qthreadpool.html, https://doc.qt.io/qt-5/qtconcurrent.html.

So, I'm still concerned, but after looking at some internal code, it seems to be common practice. :see_no_evil: Still interested what others think.

I will add some comments in a moment.

File: src/OrbitBase/include/OrbitBase/ThreadPool.h
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/ThreadPool.h:77]: Not sure if this maybe should go into the `Executor` call because ideally we don't require a thread pool. Any executor suffices. Also internally we call this function `DefaultExecutor()`, but I don't feel strongly about this.

Also it should probably return a raw pointer because it should be non-owning. This also works better with `TaskGroup` as it shouldn't own the executor.

File: src/OrbitBase/include/OrbitBase/TaskGroup.h
Comment (beckerhe) [inline src/OrbitBase/include/OrbitBase/TaskGroup.h:51]: I would prefer to keep this a non-owning pointer.

File: src/OrbitBase/ThreadPool.cpp
Comment (beckerhe) [inline src/OrbitBase/ThreadPool.cpp:220]: The thread pool should probably be created during process startup and not when `GetGlobalWorkerThreadPool` is called the first time. http://go/stadia-orbit-default-executor for how it's done internally.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM. But it would be great if someone else could also have a look.

File: src/OrbitBase/include/OrbitBase/ThreadPool.h
Comment (pierricgimmig) [reply src/OrbitBase/include/OrbitBase/ThreadPool.h]: Made it return a pointer and renamed to ThreadPool::GetDefaultThreadPool(). I left it in ThreadPool as I feel it's more descriptive and it allows to create a default thread pool in the "GetDefaultThreadPool()" method.

File: src/OrbitBase/include/OrbitBase/TaskGroup.h
Comment (pierricgimmig) [reply src/OrbitBase/include/OrbitBase/TaskGroup.h]: Done, thanks.

File: src/OrbitBase/ThreadPool.cpp
Comment (pierricgimmig) [reply src/OrbitBase/ThreadPool.cpp]: Added `void ThreadPool::InitializeDefaultThreadPool()` which is called on `OrbitApp` creation. Note that as is done internally, if we don't explicitly initialize or set the default thread pool, it will be created on the first call to `ThreadPool::GetDefaultThreadPool()`.

**Comment by pierricgimmig** [comment]
> So, I'm still concerned, but after looking at some internal code, it seems to be common practice. 🙈 Still interested what others think.
> 
> I will add some comments in a moment.

Thanks @beckerhe for your input and suggestions, I modified the PR accordingly.

File: src/OrbitBase/ThreadPool.cpp
Comment (danielfenner) [inline src/OrbitBase/ThreadPool.cpp:241]: Not sure about this: If called multiple times this invalidates pointers previouly handed out by GetDefaultThreadPool. The esiest way around this would be to just go with one default size threadpool (as suggested above).

**Comment by pierricgimmig** [comment]
> Is it actually a problem to pass `ThreadPool`s/`Executor`s to the few classes that need them?

I think the reason why there are so few occurrences of this in the code base is precisely because it's a pain to set up. I think a parallel-for loop should be trivial to use and localized in a few lines of code. Requiring to pass in a dependency to an executor for a parallel-for loop discourages its use.

**Review by danielfenner — COMMENTED** [review]
I don't have too much concerns about making the threadpool global. I think the ease of use overweights the ease of misuse.

File: src/OrbitBase/ThreadPool.cpp
Comment (pierricgimmig) [reply src/OrbitBase/ThreadPool.cpp:24]: The reason is that our interface for creating a ThreadPool returns a std::shared_ptr<ThreadPool>, we should probably change that to a std::unique_ptr actually since we can convert a std::unique_ptr to a std::shared_ptr, but not the other way around.

Comment (beckerhe) [reply src/OrbitBase/ThreadPool.cpp:24]: That's my fault. We need that shared_ptr for `TrySchedule` in `ScopedStatus` but it should really be an implementation detail of `ThreadPool` - but that's non trivial. I might be able to tackle that in the fixit week.

File: src/OrbitBase/ThreadPool.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPool.cpp:26]: I think this is just copying the behaviour that we have internally.

File: src/OrbitBase/ThreadPool.cpp
Comment (danielfenner) [reply src/OrbitBase/ThreadPool.cpp:24]: Okay, let's leave it as is for now and clean it up (or not - it's not a huge deal after all).

File: src/OrbitBase/ThreadPool.cpp
Comment (pierricgimmig) [reply src/OrbitBase/ThreadPool.cpp]: Removed "g_user_default_thread_pool". There is only "g_default_thread_pool" now, which is either one that was set by the user, or ThreadPool's internally created one. Made the interface stricter, we can still call "SetDefaultThreadPool(...)", but only once, with a non-null thread pool, and before any call to "GetDefaultThreadPool()". This makes it so that there is only ever one default thread pool created and used.

File: src/OrbitBase/ThreadPool.cpp
Comment (pierricgimmig) [reply src/OrbitBase/ThreadPool.cpp]: Implemented that approach, see comment above.

**Review by danielfenner — APPROVED** [review]
lgtm

**Pull request merged**
